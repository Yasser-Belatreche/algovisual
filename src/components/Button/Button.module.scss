.buttonContainer {
  @apply flex items-center justify-center gap-4 bg-blue text-white text-lg font-bold px-12 py-4 capitalize duration-300;
  border-radius: 10px;
  white-space: nowrap;

  &:hover {
    @apply bg-blueHover;
  }

  &.bgOrange {
    @apply bg-orange;

    &:hover {
      @apply bg-orangeHover;
    }
  }

  &.full {
    @apply w-full;
  }

  .iconContainer {
    @apply flex items-center;
    max-width: 28px;
  }

  @media only screen and (max-width: 600px) {
    @apply w-full text-base px-0;
  }
}
