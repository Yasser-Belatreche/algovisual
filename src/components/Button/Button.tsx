import React from "react";

import Image from "next/image";
import classNames from "classnames/bind";

// styles
import styles from "./Button.module.css";

const cx = classNames.bind(styles);

interface Props extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  onClick: React.MouseEventHandler<HTMLButtonElement>;
  backgroundColor: "blue" | "orange";
  icon: string;
}

/**
 * Primary UI component for user interaction
 */
const Button: React.FC<Props> = ({ children, onClick, icon, ...props }) => {
  return (
    <button onClick={onClick} {...props}>
      <p>{children}</p>
      {icon && <Image src={icon} alt="button icon" />}
    </button>
  );
};

export { Button };
