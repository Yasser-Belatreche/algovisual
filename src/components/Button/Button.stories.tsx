import React from "react";
import { ComponentStory, ComponentMeta } from "@storybook/react";

import { Button } from "./Button";
import { ICONS } from "../../utils/constants/Icons";

export default {
  title: "Global/Button",
  component: Button,
  argTypes: {},
} as ComponentMeta<typeof Button>;

const Template: ComponentStory<typeof Button> = (args) => <Button {...args} />;

export const Primary = Template.bind({});
Primary.args = { label: "start learning" };

export const Secondary = Template.bind({});
Secondary.args = {
  ...Primary.args,
  secondary: true,
};

export const FullWidth = Template.bind({});
FullWidth.args = {
  ...Primary.args,
  full: true,
};

export const ButtonWithIcon = Template.bind({});
ButtonWithIcon.args = {
  label: "contribute",
  icon: ICONS.github,
};
