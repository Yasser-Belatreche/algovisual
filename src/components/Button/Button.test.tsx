import { fireEvent, render } from "@testing-library/react";

import { Button } from "./Button";

describe("Button compoenent", () => {
  it("should have the primary color by default", () => {
    const { container } = render(<Button label="button" />);

    expect(container).toMatchSnapshot();
  });

  it("should have the secondary color when passing the secondary prop", () => {
    const { container } = render(<Button label="button" secondary />);

    expect(container).toMatchSnapshot();
  });

  it("should have the full width when passing the full prop", () => {
    const { container } = render(<Button label="button" full />);

    expect(container).toMatchSnapshot();
  });

  it("should call the onClick handler when passed", () => {
    const mockFn = jest.fn();

    const instance = render(<Button label="button" onClick={mockFn} />);

    const button = instance.getByRole("button");
    fireEvent["click"](button);

    expect(mockFn).toBeCalledTimes(1);
    expect(instance.container).toMatchSnapshot();
  });

  it("should add the class passed to props to the class list", () => {
    const instance = render(
      <Button label="button" className="myCustomClass" />
    );

    const button = instance.getByRole("button");

    expect(button.classList.length).not.toEqual(1);
    expect(button.classList).toContain("myCustomClass");
    expect(instance.container).toMatchSnapshot();
  });
});
