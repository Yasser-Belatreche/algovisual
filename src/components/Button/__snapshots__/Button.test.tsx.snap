// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button compoenent should add the class passed to props to the class list 1`] = `
<div>
  <button
    class="buttonContainer myCustomClass"
  >
    <p>
      button
    </p>
  </button>
</div>
`;

exports[`Button compoenent should call the onClick handler when passed 1`] = `
<div>
  <button
    class="buttonContainer undefined"
  >
    <p>
      button
    </p>
  </button>
</div>
`;

exports[`Button compoenent should have the full width when passing the full prop 1`] = `
<div>
  <button
    class="buttonContainer full undefined"
  >
    <p>
      button
    </p>
  </button>
</div>
`;

exports[`Button compoenent should have the primary color by default 1`] = `
<div>
  <button
    class="buttonContainer undefined"
  >
    <p>
      button
    </p>
  </button>
</div>
`;

exports[`Button compoenent should have the secondary color when passing the secondary prop 1`] = `
<div>
  <button
    class="buttonContainer bgOrange undefined"
  >
    <p>
      button
    </p>
  </button>
</div>
`;
